core/shared/src/main/scala-3/zio/internal/macros/LayerMacroUtils.scala

```diff
@@ -3,124 +3,99 @@ package zio.internal.macros
-
-    val empty: LayerExpr[E] = '{ZLayer.succeed(())}
-
-    def composeH(lhs: LayerExpr[E], rhs: LayerExpr[E]): LayerExpr[E] =
-      lhs match {
-        case '{$lhs: ZLayer[i, e, o]} =>
-          rhs match {
-            case '{$rhs: ZLayer[i2, e2, o2]} =>
-              '{$lhs.++($rhs)}
-          }
-      }
-
-    def composeV(lhs: LayerExpr[E], rhs: LayerExpr[E]): LayerExpr[E] =
-      lhs match {
-        case '{$lhs: ZLayer[i, E, o]} =>
-          rhs match {
-            case '{$rhs: ZLayer[`o`, E, o2]} =>
-              '{$lhs to $rhs}
-          }
-      }
-
-    val layerExprs: List[LayerExpr[E]] = tree.toList
-
-    ValDef.let(Symbol.spliceOwner, layerExprs.map(_.asTerm)) { idents =>
-      val exprMap = layerExprs.zip(idents).toMap
-
-      tree.fold[LayerExpr[E]](
-        empty,
-        exprMap(_).asExpr.asInstanceOf[LayerExpr[E]],
-        composeH,
-        composeV
-      ).asTerm
-
-    }.asExpr.asInstanceOf[LayerExpr[E]]
+    import quotes.reflect._
 
-  }
+    def renderExpr[A](expr: Expr[A]): String =
+      scala.util.Try(expr.asTerm.pos.sourceCode).toOption.flatten.getOrElse(expr.show)
 
-  def getNode[E: Type](layer: LayerExpr[E])(using ctx: Quotes): Node[ctx.reflect.TypeRepr, LayerExpr[E]] = {
-    import quotes.reflect._
-    layer match {
+    def getNode(layer: LayerExpr[E]): Node[TypeRepr, LayerExpr[E]] = layer match {
       case '{ $layer: ZLayer[in, e, out] } =>
         val inputs = getRequirements[in]
         val outputs = getRequirements[out]
         Node(inputs, outputs, layer)
     }
-  }
 
-  def getRequirements[T: Type](using ctx: Quotes) : List[ctx.reflect.TypeRepr] = {
-    import ctx.reflect._
+    def getRequirements[T: Type]: List[TypeRepr] = {
+      def loop(tpe: TypeRepr): List[TypeRepr] =
+        tpe.dealias.simplified match {
+          case AndType(lhs, rhs) => loop(lhs) ++ loop(rhs)
+          case AppliedType(_, TypeBounds(_,_) :: _) => Nil
+          case other if other =:= TypeRepr.of[Any] => Nil
+          case other if other.dealias.simplified != other => loop(other)
+          case other => List(other.dealias)
+        }
+
+      loop(TypeRepr.of[T])
+    }
+
+    val targetTypes = getRequirements[R]
+    val remainderTypes = getRequirements[R0]
+    val layerToDebug: PartialFunction[LayerExpr[E], ZLayer.Debug] = {
+      case '{ZLayer.Debug.tree} => ZLayer.Debug.Tree
+      case '{ZLayer.Debug.mermaid} => ZLayer.Debug.Mermaid
+    }
 
-    def loop(tpe: TypeRepr): List[TypeRepr] =
-      tpe.dealias.simplified match {
-        case AndType(lhs, rhs) =>
-          loop(lhs) ++ loop(rhs)
+    '{
+      given trace: Trace = Tracer.newTrace
+      val compose = [R1, O1, O2] => (lhs: ZLayer[R1, E, O1], rhs: ZLayer[O1, E, O2]) => lhs.to(rhs)
 
-        case AppliedType(_, TypeBounds(_,_) :: _) =>
-          List.empty
+      ${
+        def typeToNode(tpe: TypeRepr): Node[TypeRepr, LayerExpr[E]] =
+          Node(Nil, List(tpe), tpe.asType match { case '[t] => '{ZLayer.environment[t]} })
```

Comment:
Do you know whether the tracing implicit is reachable here? Shall we do this just to be safe?

```suggestion
          Node(Nil, List(tpe), tpe.asType match { case '[t] => '{ZLayer.environment[t](using trace)} })
```

Reply:
Yeah, that might be a good idea. I'm still getting used to Scala 3 macros so I'm also not really sure.